xarray/core/dataset.py

```diff
@@ -1910,6 +1910,94 @@ def from_dataframe(cls, dataframe):
             obj[name] = (dims, data)
         return obj
 
+    def to_dict(self):
+        """
+        Convert this dataset to a dictionary following xarray naming
+        conventions.
+
+        Useful for coverting to json.
+        """
+        d = {'coords': {}, 'attrs': dict(self.attrs), 'dims': dict(self.dims),
+             'data_vars': {}}
+
+        def time_check(val):
+            if np.issubdtype(val.dtype, np.datetime64):
+                val = val.astype('datetime64[us]')
+            elif np.issubdtype(val.dtype, np.timedelta64):
+                val = val.astype('timedelta64[us]')
+            return val
+
+        for k in self.coords:
+            data = time_check(self[k].values).tolist()
+            d['coords'].update({k: {'data': data,
+                                    'dims': self[k].dims,
+                                    'attrs': dict(self[k].attrs)}})
+        for k in self.data_vars:
+            data = time_check(self[k].values).tolist()
+            d['data_vars'].update({k: {'data': data,
+                                       'dims': self[k].dims,
+                                       'attrs': dict(self[k].attrs)}})
+        return d
+
+    @classmethod
+    def from_dict(cls, d):
+        """
+        Convert a dictionary into an xarray.Dataset.
+
+        Input dict can take several forms:
+
+        d = {'t': {'dims': ('t'), 'data': t},
+             'a': {'dims': ('t'), 'data': x},
+             'b': {'dims': ('t'), 'data': y}}
+
+        d = {'coords': {'t': {'dims': 't', 'data': t,
+                              'attrs': {'units':'s'}}},
+             'attrs': {'title': 'air temperature'},
+             'dims': 't',
+             'data_vars': {'a': {'dims': 't', 'data': x, },
+                           'b': {'dims': 't', 'data': y}}}
+
+        where 't' is the name of the dimesion, 'a' and 'b' are names of data
+        variables and t, x, and y are lists, numpy.arrays or pandas objects.
+
+        Parameters
+        ----------
+        d : dict, with a minimum structure of {'var_0': {'dims': [..],
+                                                         'data': [..]},
+                                               ...}
+
+        Returns
+        -------
+        obj : xarray.Dataset
+
+        See also
+        --------
+        xarray.DataArray.from_dict
+        """
+        import itertools
+
+        variables = itertools.chain(d.get('coords', {}).items(),
```

Comment:
This should be in an `else` clause of the `if` statement below, because it isn't used otherwise.